decisions/0001-bundle-granularity.rst

```diff
@@ -46,56 +46,145 @@ XBlock Directories are a client convention and not a core Blockstore primitive.
+When referencing an XBlock definition from a linked bundle, the linked bundle
+must follow the OLX Bundle conventions, but may be any type of OLX bundle:
+course, library, or other future types.
+
+The ``usage="..."`` attribute on an ``<xblock-include />`` directive in an
+XBlock definition file is considered a hint that will be used by the learning
+context (the course, library, web page, etc.) to build a "usage key" for the
+XBlock against which user state is stored. Exactly how the ``usage="..."`` hint
+is ultimately used to construct the key depends on the learning context. For
+example, an HTML block included with ``<xblock-include ... usage="bob" />`` in a
+library might have a usage key like ``lb:lib1:html:bob`` when accessed directly
+in the library, but may have a usage key like
+``block:SchoolX+course1+run:html:parent3-bob6734`` when included in a course.
+This helps with separation of concerns (bundles store definitions, which are
+OLX files; learning contexts deal with usages which are specific instantiations
+of a definition) and preserves flexibility for future learning context types.
+
+In particular, note that a learning context will almost always transform the
+usage keys of any children of an included block, in order to ensure they are
+unique. Consider this case::
+
+    <unit display_name="Example">
+        <!-- A unit in an external problem bank library, containing 3 problems -->
+        <xblock-include source="prob_bank5" definition="unit/3problems" usage="first_set" />
+        <!-- A second usage of the same problem bank -->
+        <xblock-include source="prob_bank5" definition="unit/3problems" usage="second_set" />
+    </unit>
+
+In this example, if the included ``3problems`` unit contains child includes
+like::
+
+    <xblock-include definition="problem/problem1" usage="p1" />
+    ...
+
+Then in the original example, the usage ID ``p1`` would not be unique, because
+its parent unit is included twice. So the learning context will have to prefix
+the usage hint ``p1`` in this case, producing unique usage keys like
+``first_set-p1`` for the first usage and ``second_set-p1`` for the second.
+
+---------------------------------
+Proposal: Content Library Bundles
+---------------------------------
+
+A content library bundle is an OLX bundle with additional constraints, that is
+used to represent the OLX data of a content library.
+
+As with OLX Bundles in general, the conventions defined here are not implemented
+nor enforced by Blockstore in any way, however they are documented here for the
+sake of having a central reference.
+
+A content library bundle is an OLX bundle that holds a collection of one or more
+XBlocks. Each XBlock definition (OLX ``definition.xml`` file) in the content
+library is either a "top-level block" or is a child of exactly one other XBlock
+definition in the same bundle. A content library bundle does not allow a single
+definition to have multiple different usages (multiple <xblock-include />
```

Comment:
Curios what the specific reason for this is?

Reply:
It keeps things simpler when there's 1:1 usage:definition within the content library. For example, when deleting a usage, if multiple usages are allowed, you must scan through the whole bundle to see if the definition is used anywhere else in order to know if the underlying OLX file should be deleted or left in the bundle. With this restriction, there's a much simpler process: deleting a usage deletes the definition too, unless the definition is from a linked bundle.

That said, with some of the latest refactoring that I did, it's probably not too hard to relax this restriction. We can always consider relaxing it in the future.